### PR TITLE
Clean up kafka emitter tests, add more validations and code coverage.

### PR DESCRIPTION
**Pull request by abhishekrb19**
The `KafkaEmitterTest` wasn't validating the net events emitted but only the dropped counters, which aren't particularly useful. This patch adds the missing event validation in `KafkaEmitterTest` and also cleans up the code to write unit tests easily.

Further, this module in general has insufficient code coverage overall. As a result, developers making changes in this extension often run into code coverage issues. This change should address that as well.

Additionally, the precondition config checks have been migrated to `DruidException`.

<hr>

This PR has:

- [x] been self-reviewed.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.

### CODE REVIEW

File: extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java [hunk 1]
```diff
@@ -95,12 +95,45 @@ public KafkaEmitterConfig(
...
+                          .build("druid.emitter.kafka.bootstrap.servers must be specified.");
+    }
+
+    if (this.eventTypes.contains(EventType.METRICS) && metricTopic == null) {
+      throw DruidException.forPersona(DruidException.Persona.OPERATOR)
+                          .ofCategory(DruidException.Category.NOT_FOUND)
+                          .build("druid.emitter.kafka.metric.topic must be specified"
+                                 + " if druid.emitter.kafka.event.types contains %s.", EventType.METRICS);
+    }
+    if (this.eventTypes.contains(EventType.ALERTS) && alertTopic == null) {
```
Comment (zachjsh) [inline extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java:113]: Should we combine the 3 blocks below, and includ all three of these properties in the `if druid.emitter.kafka.event.types contains %s.` part of error message? If we want to keep it separate like this, should we put that message in a static error string variable? Seems the same for each, just the type is different, which could be added using string formatter.

File: extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java [hunk 2]
```diff
@@ -95,12 +95,45 @@ public KafkaEmitterConfig(
...
-    this.segmentMetadataTopic = this.eventTypes.contains(EventType.SEGMENT_METADATA) ? Preconditions.checkNotNull(segmentMetadataTopic, "druid.emitter.kafka.segmentMetadata.topic can not be null") : null;
+
+    // Validate all required properties
+    if (bootstrapServers == null) {
+      throw DruidException.forPersona(DruidException.Persona.OPERATOR)
+                          .ofCategory(DruidException.Category.NOT_FOUND)
+                          .build("druid.emitter.kafka.bootstrap.servers must be specified.");
+    }
+
+    if (this.eventTypes.contains(EventType.METRICS) && metricTopic == null) {
```
Comment (zachjsh) [inline extensions-contrib/kafka-emitter/src/main/java/org/apache/druid/emitter/kafka/KafkaEmitterConfig.java:107]: Should we combine the 4 blocks below, and includ all three of these properties in the if druid.emitter.kafka.event.types contains %s. part of error message? If we want to keep it separate like this, should we put that message in a static error string variable? Seems the same for each, just the type is different, which could be added using string formatter.

**Comment by zachjsh** [comment]
Great addition with the new tests. Only thing I would say is would be good to add description of the new tests you added in kafkaEmitterTest; what the test is testing and what the expectations are. Great stuff!

**Review by zachjsh — APPROVED** [review]

File: extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java [hunk 3]
```diff
@@ -19,95 +19,459 @@
...
+    final Map<String, List<String>> feedToActualEvents = trackActualEventsPerFeed(eventLatch);
+
+    emitEvents(kafkaEmitter, inputEvents, eventLatch);
+
+    validateEvents(feedToExpectedEvents, feedToActualEvents);
+
+    Assert.assertEquals(0, kafkaEmitter.getMetricLostCount());
+    Assert.assertEquals(0, kafkaEmitter.getAlertLostCount());
+    Assert.assertEquals(0, kafkaEmitter.getSegmentMetadataLostCount());
+    Assert.assertEquals(0, kafkaEmitter.getSegmentMetadataLostCount());
```
Comment (TSFenwick) [inline extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java:166]: @abhishekrb19 typo error here? we have a double check on the same value

File: extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java [hunk 4]
```diff
@@ -19,95 +19,459 @@
...
```
Comment (TSFenwick) [inline extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java:210]:
```suggestion
    Assert.assertEquals(0, kafkaEmitter.getRequestLostCount()());
```

File: extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java [hunk 5]
```diff
@@ -19,95 +19,459 @@
...
```
Comment (TSFenwick) [inline extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java:253]:
```suggestion
    Assert.assertEquals(0, kafkaEmitter.getRequestLostCount()());
```

File: extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java [hunk 6]
```diff
@@ -19,95 +19,459 @@
...
```
Comment (TSFenwick) [inline extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java:346]:
```suggestion
    Assert.assertEquals(0, kafkaEmitter.getRequestLostCount()());
```

File: extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java [hunk 7]
```diff
@@ -19,95 +19,459 @@
...
```
Comment (TSFenwick) [inline extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java:389]:
```suggestion
    Assert.assertEquals(0, kafkaEmitter.getRequestLostCount()());
```

File: extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java (hunk 3 above)
Comment (abhishekrb19) [reply extensions-contrib/kafka-emitter/src/test/java/org/apache/druid/emitter/kafka/KafkaEmitterTest.java:166]: Yes, good catch. Thanks for fixing it up in https://github.com/apache/druid/pull/15890

**Pull request merged**